File: pages/login_page.py
from playwright.sync_api import Page, expect
from .base_page import BasePage


class LoginPage(BasePage):
    URL = "https://www.saucedemo.com"
    USERNAME_INPUT = "[data-test='username']"
    PASSWORD_INPUT = "[data-test='password']"
    LOGIN_BUTTON = "[data-test='login-button']"
    ERROR_MESSAGE = "[data-test='error']"

    def __init__(self, page: Page):
        super().__init__(page)

    def open(self):
        self.navigate(self.URL)
        return self

    def login(self, username: str, password: str):
        self.page.fill(self.USERNAME_INPUT, username)
        self.page.fill(self.PASSWORD_INPUT, password)
        self.page.click(self.LOGIN_BUTTON)
        return self

    def get_error_message(self) -> str:
        return self.page.text_content(self.ERROR_MESSAGE)

    def expect_error_visible(self):
        expect(self.page.locator(self.ERROR_MESSAGE)).to_be_visible()

    def expect_error_text(self, text: str):
        expect(self.page.locator(self.ERROR_MESSAGE)).to_contain_text(text)

File: src/pages/LoginPage.ts
import { Page, expect } from '@playwright/test';
import { BasePage } from './BasePage';
import { BASE_URL } from '@utils/config';

export class LoginPage extends BasePage {
  private readonly usernameInput = "[data-test='username']";
  private readonly passwordInput = "[data-test='password']";
  private readonly loginButton = "[data-test='login-button']";
  private readonly errorMessage = "[data-test='error']";

  constructor(page: Page) {
    super(page);
  }

  async open(): Promise<this> {
    await this.navigate(BASE_URL);
    return this;
  }

  async login(username: string, password: string): Promise<this> {
    await this.page.fill(this.usernameInput, username);
    await this.page.fill(this.passwordInput, password);
    await this.page.click(this.loginButton);
    await this.pause();
    return this;
  }

  async getErrorMessage(): Promise<string | null> {
    return this.page.textContent(this.errorMessage);
  }

  async expectErrorVisible(): Promise<void> {
    await expect(this.page.locator(this.errorMessage)).toBeVisible();
  }

  async expectErrorText(text: string): Promise<void> {
    await expect(this.page.locator(this.errorMessage)).toContainText(text);
  }
}
